runtime/domain.c

```diff
@@ -2096,6 +2127,105 @@ static void domain_terminate (void)
+
+  caml_finish_major_cycle(0);
+
+  caml_finish_marking();
+
+  caml_orphan_ephemerons(domain_state);
+  caml_orphan_finalisers(domain_state);
+
+  /* domain_state may be re-used by a fresh domain here (now that we
+   * have done remove_from_stw_domains and released the
+   * all_domains_lock). However, domain_create() won't touch it until
+   * it has claimed the domain_lock, so we hang onto that while we are
+   * tearing down the state. */
+
+  /* Delete the domain state from statmemprof after any promotion
+   * (etc) done by this domain: any remaining memprof state will be
+   * handed over to surviving domains. */
+  caml_memprof_delete_domain(domain_state);
+
+  /* We can not touch domain_self->interruptor after here
+     because it may be reused */
+  caml_remove_generational_global_root(&domain_state->dls_root);
+  caml_remove_generational_global_root(&domain_state->backtrace_last_exn);
+  caml_stat_free(domain_state->final_info);
+  caml_stat_free(domain_state->ephe_info);
+  caml_free_intern_state();
+  caml_free_extern_state();
+  /* FIXME: this fails the caml_mark_stack_is_empty assert in the debug runtime
+    caml_teardown_major_gc();
+   */
+
+  caml_finalise_heap();
+
+  caml_teardown_shared_heap(domain_state->shared_heap);
+  domain_state->shared_heap = 0;
+  caml_free_minor_tables(domain_state->minor_tables);
+  domain_state->minor_tables = 0;
+
+  caml_orphan_alloc_stats(domain_state);
+  /* Heap stats were orphaned by [caml_teardown_shared_heap] above.
+     At this point, the stats of the domain must be empty.
+
+     The sampled copy was also cleared by the minor collection(s)
+     performed above at [caml_empty_minor_heaps_once()], see the
+     termination-specific logic in [caml_collect_gc_stats_sample_stw].
+  */
+
+  /* TODO: can this ever be NULL? can we remove this check? */
+  if(domain_state->current_stack != NULL) {
+    caml_free_stack(domain_state->current_stack);
+  }
+  caml_free_backtrace_buffer(domain_state->backtrace_buffer);
+  caml_free_gc_regs_buckets(domain_state->gc_regs_buckets);
+}
+
+void caml_stop_all_domains(void)
+{
+  atomic_store_relaxed(&domains_exiting, 1);
+
+  /* The following is similar to caml_interurpt_all_signal_safe(), but
+     ignoring the currently running domain. */
+  for (dom_internal *d = all_domains; d < &all_domains[Max_domains]; d++) {
+    if (d->state == Caml_state)
+      continue;
+    atomic_uintnat *interrupt_word =
+      atomic_load_acquire(&d->interruptor.interrupt_word);
+    /* Early exit: if the current domain was never initialized, then
+       neither have been any of the remaining ones. */
+    if (interrupt_word == NULL)
+      break;
+    atomic_store_relaxed(&d->force_exit, 1);
+    interrupt_domain(&d->interruptor);
+  }
+
+  /* Give the threads some time to terminate */
+  int i = atomic_load_acquire(&caml_num_domains_running) + 5;
+  while (!caml_domain_alone() && i != 0) {
+#ifdef WIN32
+    Sleep(1);
```

Comment:
This sleep-based synchronization looks very fishy to me -- but then I'm not a domain expert. Is there not a standard synchronization primitive that we can use?

Reply:
It's only ugly (-:

If we reach this state, there is absolutely no guarantee that the interrupt request will be processed by the domains, so we can not wait for them without an upper time bound, in case at least one domain is unresponsive. The time limit here is (number of currently running domains + 5) x 1 millisecond. That's an arbitrary number which hopefully allows every non-sleeping domain to be scheduled and receive the exit request.

Reply:
I understand the concern, but I am not sure what scenarios you have in mind about delayed processing.

I am assuming that all domains have at least a thread that checks for interruptions on a regular basis -- otherwise this is a bug that should be fixed to restore this property, because the minor GC relies on it. I am not sure if you make a different assumption, or if you also assume this but then are thinking of the *other* threads on that same domain that are waiting on other blocking calls, or something else?

As long as the owner of the domain lock regularly handles interruptions, we know (I think?) that the other threads on the domain will not use the runtime again before they take the domain lock. So we could have a check for "we are forced to terminate" at that point that would shut them down before they return to OCaml code, and not have to wait at all for those threads right now.